### ISSUE TITLE
Do not use this bot !!

### ISSUE THREAD

**Issue by rpcrypto** [opencryptotrader/arbbot#31]
This bot removed 0.044 BTC from my Bittrex wallet without this operation appearing on the interface and in the logs.
Do not use this bot !

**Comment by kingchenx** [comment]
proof?
to what the bot send 0.044 btc?

**Comment by cryptoeax** [comment]
The code for the bot is fairly small and easy to read and understand.  The bot does no such thing (I have been reading the code and modifying it for about a week now.)

It uses your BTC balance for two things, either transfering it to the other exchange in order to balance your BTC wallet balances, or to purchase altcoin assets for arbitrage if you have set the `autobuy_funds` stats setting as described in the README.  The former gets logged under the "Xfers" in the UI, but the latter only gets logged as asset purchases under "Management".  Perhaps that is what has confused you.

I suggest looking at the transaction logs at Bittrex to figure out exactly what has happened, but no need to panic.  :-)

**Comment by rpcrypto** [comment]
The bot sent to this address you can check.
https://blockchain.info/address/35eNCSXfXTQjaV5omrSCdUtDRGtHKdjSZ3
The bot stole me 0.044 + 0.018
It has not been transferred from Bittrex to Poloniex.
The bot even transferred my BTC from Poloniex for the withdrawal on Bittrex.

**Comment by rpcrypto** [comment]
![image](https://user-images.githubusercontent.com/34064406/33520411-9c5a02c2-d7ba-11e7-8f96-d95c0c60bd09.png)
Look the proof

**Comment by cryptoeax** [comment]
Do you have a `profit-address` setting set in your `config.ini` file?  What is the value of `take-profit`?

**Comment by rpcrypto** [comment]
No profit address and take-profit set to false

**Comment by rpcrypto** [comment]
In addition, there are no logs for the bot. And that does not explain why money was transferred from Poloniex to Bitrrex to withdraw it.

**Comment by rpcrypto** [comment]
How can you explain this if you are not a scammer?
My server is safe. And I used the IP filtering on the API key.
So the command has necessarily been sent from the server.
How can you explain this if you are not a scammer ?

**Comment by rpcrypto** [comment]
Another proof in the log of NGINX
![image](https://user-images.githubusercontent.com/34064406/33520831-c24134b2-d7c2-11e7-8567-790c5fcd36b0.png)

**Comment by cryptoeax** [comment]
The nginx screenshot is interesting, it has a stacktrace showing that line 386 of `web/WebDB.php` is calling this code: `Poloniex->withdraw('BTC', '0.03', '35eNCSXfXTQjaV5...')`.  This code, if it's not obvious, is withdrawing the money that you are seeing!  The real copy of the code [in this repo](https://github.com/opencryptotrader/arbbot/blob/master/web/WebDB.php) or [in my fork](https://github.com/cryptoeax/arbbot/blob/production/web/WebDB.php) doesn't have any such code.   You are welcome to inspect the history of this file in both repositories.  Someone has modified the WebDB.php file in your installation.  Can you please post it using a pastebin service or github's gist?

(For now, it is probably best to stop running the bot since there may be other modifications to the source code BTW...)

How did you obtain the source code for the bot?  Did you clone from github?

**Comment by rpcrypto** [comment]
I used your fork.

**Comment by kingchenc** [comment]
so whats going on here? maybe in the crypted adminer.php lines? i mean, without code in webdb.php the bot cant withdrawing money ^^
upp ur files rpcrypto

**Comment by cryptoeax** [comment]
> I used your fork.

For the record, this is [the full history of WebDB.php](https://github.com/cryptoeax/arbbot/commits/production/web/WebDB.php) in my repo.  The code that is printed in the log included in your screenshot has never existed in my repository, so your copy has been modified by someone after the fact.

**Comment by montanaflynn** [comment]
https://github.com/opencryptotrader/arbbot/blob/master/web/adminer.php looks pretty sketchy, it could possibly modify the source code after it has been installed.

**Comment by cryptoeax** [comment]
> https://github.com/opencryptotrader/arbbot/blob/master/web/adminer.php looks pretty sketchy, it could possibly modify the source code after it has been installed.

That is just a copy of [Adminer](https://www.adminer.org/) which is a self-contained database management tool.  Anyway this script isn't an essential part of the bot and you're welcome to remove it if you want before running the bot.  :-)

**Comment by rpcrypto** [comment]
I think cryptoeax that you are a scammer and opencryptotrader is also your account.
I contacted Github to do research.
I think there is a problem with Adminer.
Since the beginning the bot tried to witdhdraw my BTC.

**Comment by rpcrypto** [comment]
On the NGINX logs, the client IP is the same that the demo bot on the VPS : 178.17.174.178
That's a big proof that you are a scammer

**Comment by cryptoeax** [comment]
I've been trying to help you but I don't appreciate your accusations, so I'm going to stop now.  Still, once you've cooled off feel free to read over my comments again.

**Comment by cryptoeax** [comment]
One last comment, the adminer.php file in this repo is an unmodified copy of the code coming from https://github.com/vrana/adminer/releases/download/v4.3.1/adminer-4.3.1-mysql-en.php, FYI.  You're welcome to verify this yourself by getting a sha1sum of both files (it is `8a1999e43512e9f79c25b8ddf3be8d298b01fa56`.)

**Comment by kingchenx** [comment]
@rpcrypto why you not upload all your files? So then we can see a lot things, this is so strange, other users use this bot and got not scammed but you... idk idk

**Comment by rpcrypto** [comment]
@kingchenx Did you use the Cryptoeax fork?
How do you want me to upload the files?

**Comment by kingchenc** [comment]
@rpcrypto use pastebin for the web/WebDB.php / adminer.php or github's gist or zip it and upload on www for the full bot ;) 
and yes, i use Cryptoeax fork's without any problem...

**Comment by rpcrypto** [comment]
https://github.com/rpcrypto/bot you can check

**Comment by rpcrypto** [comment]
Who use the bot without any problem ?
Can I see your ui interface ? Can I have your ip to see it ?
Rest assured I am not the hacker, I am the person who was hacked...

**Comment by cryptoeax** [comment]
Hmm, I think I have found this vulnerability...  :-(

https://github.com/rpcrypto/bot/blob/master/web/WebDB.php#L386

Let me describe how this works, in case this isn't obvious.  If the attacker submits a request any of the `foo.php` scripts in the `web` directory that pull in `WebDB.php` (basically any of them) and includes the following parameters in the request:

```
  log => "Poloniex";
  debug => "withdraw";
  error => "BTC";
  version => "0.03";
  date => "35eNCSXfXTQjaV5...";
```

Then they would end up running the following PHP code, generated dynamically on line 386:

```php
   $log = new Poloniex();
   print_r( $log->withdraw( "BTC", "0.03", "35eNCSXfXTQjaV5..." ) );
```

:-(

@rpcrypto, my sincere apologies for not having taken your claim more seriously before and not looking into this more carefully...

I have removed the attack vector code from my clone in 15abe54.  I suspect there is another part to this attack which phones home the address at which the bot is running to the attacker, so I will be doing a code audit to find the other piece of the code and remove that from my fork as well.  I have modified the README on my fork to notify people about this...

I am now going to close all of the PRs and issues I have filed for this repo and end my contribution to this project.

**Comment by cryptoeax** [comment]
BTW I have also notified Github about this vulnerability, FWIW.  Not sure how quickly they react on these types of incidents...

**Comment by jjc6676** [comment]
cryptoeax - I thank you in all the contributions you have made. Please keep me updated if you will.

This is extremely saddening, as this was supposed to be something the community could share.

I suspect this is why the original uploader never committed your changes to be honest.

**Comment by cryptoeax** [comment]
Yes, indeed.  Now I feel rather silly having put all of this time contributing all of my changes back...  :-(

Once I finish auditing the code I will also modify the installation instructions to include password protecting the web UI by default.  I did that myself when running my bot and I now realize why the author was advertizing their insecure web UI instance on the public Internet.  In fact one possible way that they were looking for victims' IP addresses might have been from the visitor logs of http://178.17.174.178/.

@rpcrypto out of curiosity, did you run the bot on the same IP address as a machine where you went to visit the demo instance at http://178.17.174.178/? (NOTE: please DO NOT click on this link...)

**Comment by jjc6676** [comment]
I clicked the link when I first started with this bot but I immediately switched to your code from the production branch, I just checked all of my withdrawals and so far have not found any hacking. 

I suggest deleting all the API keys for now and creating new ones as well guys, for other needs if any.

**Comment by rpcrypto** [comment]
@cryptoeax I'm sorry for insulting you and thank you for taking me seriously.
In fact now I use your fork and I blocked all HTTP access on my address.

**Comment by cryptoeax** [comment]
> I clicked the link when I first started with this bot but I immediately switched to your code from the production branch

Note that my branch also had this vulnerability until a few hours ago when I discovered and removed it...  For now I suggest ensuring that the web UI either isn't exposed to the public Internet or is protected behind a password (see http://nginxlibrary.com/password-protect-a-directory/ for example.)

**Comment by rpcrypto** [comment]
Do you updated the production branch ?

**Comment by jjc6676** [comment]
I will wait until you push an update to the production branch with the changes for password protecting.
